File: Algorithms/src/1404. Number of Steps to Reduce a Number in Binary Representation to One/Code/solution.cpp
class Solution
{
public:
    int numSteps(string s)
    {
        int steps = 0;
        int carry = 0;

        // Traverse from right to left (ignore first bit for now)
        for (int i = s.size() - 1; i > 0; i--)
        {
            int bit = (s[i] - '0') + carry;

            if (bit == 1)
            {
                // Odd case: add 1 (carry becomes 1), then divide
                steps += 2;
                carry = 1;
            }
            else
            {
                // Even case: just divide by 2
                steps += 1;
                // carry remains same (if bit was 2, carry stays 1)
            }
        }

        // If carry remains, we need one extra step
        return steps + carry;
    }
};

File: Algorithms/src/1404. Number of Steps to Reduce a Number in Binary Representation to One/Code/solution.py
class Solution:
    def numSteps(self, s: str) -> int:
        steps = 0
        carry = 0
        
        # Traverse from right to left (ignore first bit)
        for i in range(len(s) - 1, 0, -1):
            bit = int(s[i]) + carry
            
            if bit == 1:
                # Odd case
                steps += 2
                carry = 1
            else:
                # Even case
                steps += 1
        
        return steps + carry